qiskit/utils/deprecation.py

```diff
@@ -89,27 +102,96 @@ def wrapper(*args, **kwargs):
-                    "will in future be removed.",
-                    category=category,
-                    stacklevel=3,
-                )
-            else:
-                warnings.warn(
-                    f"{func_name} keyword argument {old_arg} is deprecated and "
-                    f"replaced with {new_arg}.",
-                    category=category,
-                    stacklevel=3,
-                )
-
-                kwargs[new_arg] = kwargs.pop(old_arg)
+        if old_arg not in kwargs:
+            continue
+        if new_arg in kwargs:
+            raise TypeError(f"{func_name} received both {new_arg} and {old_arg} (deprecated).")
+        warnings.warn(old_kwarg_to_msg[old_arg], category=category, stacklevel=3)
+        if new_arg is not None:
+            kwargs[new_arg] = kwargs.pop(old_arg)
+
+
+def _add_deprecation_to_docstring(
+    func: Callable, msg: str, *, since: Optional[str], pending: bool
+) -> None:
+    if "\n" in msg:
+        raise ValueError(
+            "Deprecation messages cannot contain new lines (`\\n`), but the deprecation for "
+            f'{func.__qualname__} had them. Usually this happens when using `"""` multiline '
+            f"strings; instead, use string concatenation.\n\n"
+            "This is a simplification to facilitate deprecation messages being added to our "
+            "documentation. If you have a compelling reason to need "
+            "new lines, feel free to improve this function or open a request at "
+            "https://github.com/Qiskit/qiskit-terra/issues."
+        )
+
+    if since is None:
+        version_str = "unknown"
+    else:
+        version_str = f"{since}_pending" if pending else since
+
+    # We insert deprecations in-between the description and the meta sections. The below is from
+    # https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections.
+    # We only include what is used by Qiskit projects for functions/methods to avoid wasting
+    # computation.
+    meta_lines = {
+        "Args:",
+        "Arguments:",
+        "Example:",
+        "Examples:",
+        "Note:",
+        "Notes:",
+        "Parameters:",
+        "Return:",
+        "Returns:",
+        "Raises:",
+        "References:",
+        "Yield:",
+        "Yields:",
+    }
+
+    indent = ""
+    meta_index = None
+    if func.__doc__:
+        original_lines = func.__doc__.splitlines()
+        content_encountered = False
+        for i, line in enumerate(original_lines):
+            stripped = line.strip()
+
+            # Determine the indent based on the first line with content. But, we don't consider the
+            # first line, which corresponds to the format """Docstring.""", as it does not properly
+            # capture the indentation of lines beneath it.
+            if not content_encountered and i != 0 and stripped:
+                num_spaces = len(line) - len(line.lstrip())
+                indent = " " * num_spaces
```

Comment:
Can you also add a test case for this?

Reply:
Oh, good idea! Thanks. And I like the variable name!

Reply:
Actually I don't think this matters with using `' '`. `.splitlines()` will already have dealt with `\n` by treating them as separate lines. I don't expect anyone to be using `\t` or `\r`. This may be "premature generalization", where we make things more robust than they need to be, which has a cost to understanding the code.

Sphinx is pretty aggressive about erroring when things are broken, like bad indentation, so we can always fix this is if I'm wrong and we do encounter the edge case in the real world.

But, I like the variable rename.